test/unit/puppetlabs/puppetserver/lockable_pool_test.clj

```diff
@@ -167,54 +207,332 @@
     (let [pool (create-populated-pool 2)]
       (.lock pool)
       (is (true? true))
-      (let [borrow-thread-1 (future (let [instance (.borrowItem pool)]
+      (let [borrow-thread-started-1? (promise)
+            borrow-thread-started-2? (promise)
+            borrow-thread-borrowed-1? (promise)
+            borrow-thread-borrowed-2? (promise)
+            borrow-thread-1 (future (deliver borrow-thread-started-1? true)
+                                    (let [instance (.borrowItem pool)]
+                                      (deliver borrow-thread-borrowed-1? true)
                                       (.releaseItem pool instance)))
-            borrow-thread-2 (future (let [instance (.borrowItem pool)]
+            borrow-thread-2 (future (deliver borrow-thread-started-2? true)
+                                    (let [instance (.borrowItem pool)]
+                                      (deliver borrow-thread-borrowed-2? true)
                                       (.releaseItem pool instance)))]
-        ;; this is racey, but the only ways i could think of to make it non-racey
-        ;; depended on knowledge of the implementation
-        ;; Uncomment and make non-racey when we have an implementation that
-        ;; allows reentrant borrows.
-        ;; (Thread/sleep 500)
-        ;; (is (not (realized? borrow-thread-1)))
-        ;; (is (not (realized? borrow-thread-2)))
-
-        ;; Current implementation of JRubyPool is not reentrant. This assertion
-        ;; tests that the error we expect is thrown, rather than a deadlock
-        ;; happening. The tests for the behavior we want have been commented out
-        ;; until we change the implementation.
-        (is (thrown? IllegalStateException (.borrowItem pool)))
-        ;; (let [instance (.borrowItem pool)]
-        ;;   (is (true? true))
-        ;;   (.releaseItem pool instance))
+        @borrow-thread-started-1?
+        @borrow-thread-started-2?
+        (is (not (realized? borrow-thread-borrowed-1?)))
+        (is (not (realized? borrow-thread-borrowed-2?)))
+
+        (let [instance (.borrowItem pool)]
+          (is (true? true))
+          (.releaseItem pool instance))
 
         (is (true? true))
         (.unlock pool)
         @borrow-thread-1
         @borrow-thread-2))))
 
+(deftest pool-lock-reentrant-for-one-lock-test
```

Comment:
I can eliminate this one.  I sometimes like to have a discrete "1 instance" vs. a "many instance" test since there might be a problem in the implementation which is brought out in one case but not the other.  I didn't have any specific code paths in mind in this situation, though.